arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/FluxK.kt

```diff
@@ -91,32 +94,48 @@ data class FluxK<out A>(val flux: Flux<out A>) : FluxKOf<A> {
    *  ```
    */
   fun <B> bracketCase(use: (A) -> FluxKOf<B>, release: (A, ExitCase<Throwable>) -> FluxKOf<Unit>): FluxK<B> =
-    FluxK(Flux.create<B> { sink ->
-      flux.subscribe({ a ->
-        if (sink.isCancelled) release(a, ExitCase.Cancelled).fix().flux.subscribe({}, sink::error)
-        else try {
-          sink.onDispose(use(a).fix()
-            .flatMap { b -> release(a, ExitCase.Completed).fix().map { b } }
-            .handleErrorWith { e -> release(a, ExitCase.Error(e)).fix().flatMap { FluxK.raiseError<B>(e) } }
-            .flux
-            .doOnCancel { release(a, ExitCase.Cancelled).fix().flux.subscribe({}, sink::error) }
-            .subscribe({ sink.next(it) }, sink::error, { }, {
-              sink.onRequest(it::request)
-            })
+    Flux.create<B> { emitter ->
+      val acquireActive = AtomicBooleanW(true)
+      val cancelable = Disposables.composite()
+      emitter.onCancel(cancelable)
+
+      value().subscribe(
+        { a: A ->
+          cancelable.add(
+            Flux.defer { use(a).value() }
+              .doOnCancel { Flux.defer { release(a, ExitCase.Cancelled).value() }.subscribe() }
+              .subscribe({ b ->
+                Flux.defer { release(a, ExitCase.Completed).value() }
+                  .subscribe({ emitter.next(b) }, { e -> emitter.error(e) })
+              }, { e ->
+                Flux.defer { release(a, ExitCase.Error(e)).value() }
+                  .subscribe({ emitter.error(e) }, { e2 -> emitter.error(Platform.composeErrors(e, e2)) })
+              }, {
+                if (!acquireActive.value) emitter.complete()
```

Comment:
The above meaning that we could potentially check this within a test if it makes sense :D

Reply:
Yes, you're right and it can be worse.

```
val acquire = Flux.just(10, 9, 8, 7, 6, 5, 4, 3, 2, 1).k()
val use = { i: Int -> Flux.timer(i, TimeUnit.SECONDS).map { i }.k() }
val release = { _, _ -> FluxK { println("Release $it") } }

acquire.bracketCase(release, use)
```

- All these timers will start running in parallel since we call `use` immediately in `onNext`
- The first element takes the longest to finish, so when `acquire` completes and `acquireActive` is set to false, and `Flux.timer(1, TimeUnit.SECONDS)` will finish before all other `use` calls are finished.

In other words the `emitter` is closed and the other values from `use` are never observed since they're dropped.

Reply:
Maybe instead of an AtomicBooleanW we can have an atomic of a list(?) and let onNext's push something (maybe just A) and onComplete push something different (Unit?), so when onNext's `use` finishes we pop onNext's own element and check for the one from onComplete in order to complete 🤔